core/sequencer/election/tracker.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/monitoring"
+	"github.com/google/trillian/util/election2"
+)
+
+const resourceLabel = "resource"
+
+var (
+	once     sync.Once
+	isMaster monitoring.Gauge
+)
+
+func createMetrics(mf monitoring.MetricFactory) {
+	isMaster = mf.NewGauge(
+		"is_master",
+		"Set to 1 for resources for which this instance is currently master",
+		resourceLabel)
+}
+
+type mastership struct {
+	e        election2.Election
+	acquired time.Time
+}
+
+// Tracker tracks mastership of a collection of resources.
+type Tracker struct {
+	factory     election2.Factory
+	maxHold     time.Duration
+	master      map[string]mastership
+	masterMu    sync.RWMutex
+	watching    map[string]bool
+	watchingMu  sync.RWMutex
+	newResource chan string
+}
+
+// NewTracker returns a new mastership tracker.
+func NewTracker(factory election2.Factory, maxHold time.Duration, metricFactory monitoring.MetricFactory) *Tracker {
+	once.Do(func() { createMetrics(metricFactory) })
+	return &Tracker{
+		factory:     factory,
+		maxHold:     maxHold,
+		master:      make(map[string]mastership),
+		watching:    make(map[string]bool),
+		newResource: make(chan string),
+	}
+}
+
+// AddResource makes the mastership tracker aware of new resources.
+// The same resource may be added an unlimited number of times.
+func (mt *Tracker) AddResource(res string) {
+	mt.newResource <- res
+}
+
+// Run starts new watchers for new resources.
+func (mt *Tracker) Run(ctx context.Context) {
+	for {
+		select {
+		case res := <-mt.newResource:
+			if mt.setWatching(res) {
+				go func() {
+					defer mt.setNotWatching(res)
+					if err := mt.watchResource(ctx, res); err != nil {
+						glog.Errorf("watchResource(%v): %v", res, err)
+					}
+				}()
+			}
+		case <-ctx.Done():
+			glog.Infof("election: Run() exiting due to expired context: %v", ctx.Err())
+			return
+		}
+	}
+}
+
+// watchResource is a blocking method that runs elections for res and updates mt.master.
+func (mt *Tracker) watchResource(ctx context.Context, res string) error {
+	e, err := mt.factory.NewElection(ctx, res)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := e.Close(ctx); err != nil {
```

Comment:
Is there a condition under which this call can race with `WithMastership` in line 194? Looks okay to me, as `watchOnce` returns errors only when the election object is not visible to line 194, so should be no race.

Reply:
Hm, unless the loop 107 stops because `ctx` is canceled...

Reply:
`defer mt.setNotMaster()` in `watchOnce` will always be called before `e.Close()`.